src/index.ts

```diff
@@ -136,21 +137,32 @@ const attempt = async <T, E extends Error>(
 
 export const go = async <T, E extends Error>(
   fn: () => Promise<T>,
-  options?: GoAsyncOptions
+  options?: GoAsyncOptions<E>
 ): Promise<GoResult<T, E>> => {
   if (!options) return attempt(fn);
 
-  const { retries, attemptTimeoutMs, delay, totalTimeoutMs } = options;
+  const { retries, attemptTimeoutMs, delay, totalTimeoutMs, afterFailedAttempt } = options;
 
-  let fullTimeoutExceeded = false;
+  let fullTimeoutExceededGoResult: GoResult<never, Error> | null = null;
   let totalTimeoutCancellable: CancellableTimeout | null = null;
   let fullTimeoutPromise = new Promise((_resolve) => {}); // Never resolves
   if (totalTimeoutMs !== undefined) {
     // Start a "full" timeout that will stop all retries after it is exceeded
     totalTimeoutCancellable = cancellableSleep(totalTimeoutMs);
     fullTimeoutPromise = totalTimeoutCancellable.promise.then(() => {
-      fullTimeoutExceeded = true;
-      return fail(new Error('Full timeout exceeded'));
+      const goRes = fail(new Error('Full timeout exceeded'));
+      fullTimeoutExceededGoResult = goRes;
+
+      // Ignore the result of afterFailedAttempt callback
+      //
+      // NOTE: The casting is technically not valid (because user might expect a custom E type and we use regular Error
+      // class), but the assumption is that when users provide explicit types they expect the timeout will not happen.
+      //
+      // Alternatively, we could type the result as "GoResult<T, E | Error>" but users would need to explicitely check
+      // if the error is the class they expected (which we do not want).
+      if (afterFailedAttempt) goSync(() => afterFailedAttempt(goRes as GoResultError<E>));
+
+      return goRes;
```

Comment:
Do we even want to run `afterFailedAttempt` callback for the full timeout? Because technically it's not a failed attempt. And what you're passing to `afterFailedAttempt` is the result of `go`. That will lead only to duplicit logs IMO.

Reply:
I can imagine use cases for both. I think it's also reasonable to call the callback after an attempt which is timeouted. It is a bit strange that `go(fn, {totalTimeoutMs: 1000})` would not trigger the `afterFailedAttempt ` (it would seem like there was no attempt at all?).

Also, the `go(fn, {attemptTimeoutMs: 100})` and `go(fn, {totalTimeoutMs: 100})` should imo behave the same so I think we should either do both your suggestions or none.

Reply:
Ok, I'll do it :D 

But I still think it will be a bit confusing. In the worst case we will change it back (since we are the only user of this library anyway :D ).

Reply:
To me, I would think that `afterFailedAttempt` should not fire if the total timeout is reached. It looks like this is the current behaviour you've got now (?)

Maybe a better name would be `onAttemptError` since that is a bit clearer it's only called when an attempt promise rejects, since "failed" is maybe a bit ambiguous? Also I think "on..." is a slightly more standard callback name

Reply:
> To me, I would think that afterFailedAttempt should not fire if the total timeout is reached. It looks like this is the current behaviour you've got now (?)

Yeah, only failed attempts trigger this callback as of now. But this is not fired for the final go result.

> Maybe a better name would be onAttemptError since that is a bit clearer it's only called when an attempt promise rejects, since "failed" is maybe a bit ambiguous? Also I think "on..." is a slightly more standard callback name

The name sounds better, I'll update. Btw. this will not fire for the last unsuccessful attempt (the one returned by go).